Cache the embeddings when requested

orionw opened the pull request:
<!-- If you are submitting a dataset or a model for the model registry please use the corresponding checklists below otherwise feel free to remove them. -->

<!-- add additional description, question etc. related to the new dataset -->
This PR would solve #946 and allow the user to cache embeddings with:
```
# define your task and model above as normal
...
# wrap the model with the cache wrapper
from mteb.models.cache_wrapper import CachedEmbeddingWrapper
model_with_cached_emb = CachedEmbeddingWrapper(model, cache_path='path_to_cache_dir')
# run as normal
evaluation.run(model, ...) 
```

I have it right now set to cache the embeddings per task, but this could also be a general cache which could be used across tasks which share documents (like Wikipedia, etc.). I can modify it to have a general cache, but there are perks to the task-specific cache also (like being able to delete just one portion). 

I tried to make it scalable, although I haven't stress tested large datasets yet. 

## Checklist
<!-- Please do not delete this -->

- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`.

Samoed commented:
I created a wrapper for the `SentenceTransformer` model in #1278. I believe it might be possible to combine them in some way

orionw commented:
Ah shoot, sorry I missed that @Samoed! Where exactly is it?

Samoed commented:
https://github.com/embeddings-benchmark/mteb/blob/c2cf927461bc53395c1eab83acb4f11472934633/mteb/models/sentence_transformer_wrapper.py

orionw commented:
Gotcha, thank you @Samoed! I think, and please correct me if I'm wrong, that your wrapper handles the prompting.  I think my wrapper is only for caching and only happens on demand (shouldn't be a default).  I think they can co-exist, but let me know if I am misunderstanding!

orionw commented:
> Also, I've updated the API of the Encoder interface, and it no longer includes encode_queries and encode_corpus.

Ah, good to know! What has replaced it? Happy to modify this to wrap the function that needs to be wrapped.  The core of this is just wrapping the main encode function so that we can create a map of text->embedding.

Samoed commented:
In my PR, the wrapper has:
```python
def encode(
        self,
        sentences: Sequence[str],
        *,
        task_name: str,
        prompt_type: PromptType | None = None,
        **kwargs: Any,
    ) -> np.ndarray:
    ....
    return self.model.encode(
            sentences,
            prompt_name=prompt_name,
            **kwargs,
        )
```
`task_name` is used to get the prompt relevant to the task, and `prompt_type` specifies the type of prompt (e.g., query or passage). However, it's currently difficult to determine which wrapper will be applied first and which parameters should be supported.

Samoed commented on README.md:
````diff
@@ -369,6 +369,21 @@ df = results_to_dataframe(results)
+
+```python
+... # define your task and model above as normal
+evaluation.run(model, encode_kwargs={'cache_embeddings': 'path_to_cache_dir', "vector_dim": 512}) # setting this flag will create/use the cache
````
Maybe make this 
```python
evaluation.run(model, cache_embeddings='path_to_cache_dir', vector_dim= 512)
```
Also, is it possible to make it without specifying vector_dim?

orionw replied:
+1, you can not set it it and it will default to 768.  But to create the mmap matrix it needs some dimension.

I could have it wait until it encodes something and then get the dimension/create the matrix? And/or store a dimensionality number in the cache so it could be loaded without it needing to be passed. Does that seem preferable?

Samoed replied:
I see that you passed default, but what will be if model has 1024 vector size? Will it be loaded correctly? I think that load from cache will be more preferable

orionw replied:
Sounds good to me, added the inferring details in so it doesn't have to be passed.

orionw commented on mteb/evaluation/CacheWrapper.py:
```diff
@@ -0,0 +1,261 @@
+
+        if hasattr(model, "encode_queries") and hasattr(model, "encode_corpus"):
+            self.encode_method = "split"
+            self.query_cache = TextVectorMap(
```
Agree! I do that here: https://github.com/embeddings-benchmark/mteb/blob/89db68e943713414e2eca1680085572a8c0790e8/mteb/evaluation/MTEB.py#L378 which is then passed to this class.

Samoed replied:
Ah, I see. But since you're recreating the model, this would result in a wrapper inside another wrapper, leading to multiple caches of the same data.

Samoed replied:
In main branch `task_name` and `prompt_type` currently passing to model_encode and from there to `model.encode` in kwargs in `prompt_name`

orionw replied:
Thanks @Samoed for pointing out this issue.  I've moved it out to avoid this issue!

Samoed commented:
Yes, I've made changes so that parameters are passed to all wrappers

KennethEnevoldsen reviewed (commented):
Looking great @orionw! Made a few suggestions. Wrapping the SentenceTransformerWrapper should not be a problem at all.

I see that you use a pkl as a wrapper, I was thinking it might be better to use a data format that is [safe](https://www.benfrederickson.com/dont-pickle-your-data/). 
Mostly if people intended to share this data.

orionw commented on README.md:
````diff
@@ -369,6 +369,21 @@ df = results_to_dataframe(results)
+```python
+... # define your task and model above as normal
+evaluation.run(model, encode_kwargs={'cache_embeddings': 'path_to_cache_dir'}) # setting this flag will create/use the cache
+```
````
+1 great feedback, thanks @KennethEnevoldsen! I suppose if it scales too big, we can always switch to ujson or similar. 

Update it to work like this, agree it's much easier to maintain like that.

orionw commented:
> I see that you use a pkl as a wrapper, I was thinking it might be better to use a data format that is [safe](https://www.benfrederickson.com/dont-pickle-your-data/).

I went with a Pickle for the compression and speed, but agree JSON is better overall. I switched back to it and moved it to the models section of the code.

KennethEnevoldsen reviewed (approved):
I believe things are looking ready to merge.

KennethEnevoldsen commented on mteb/models/cache_wrapper.py:
```diff
@@ -0,0 +1,330 @@
+
+
+class CachedEmbeddingWrapper:
+    def __init__(self, model: Any, cache_path: Union[str, Path]):
```
```suggestion
    def __init__(self, model: Encoder, cache_path: Union[str, Path]):
```

KennethEnevoldsen commented on mteb/models/cache_wrapper.py:
```diff
@@ -0,0 +1,330 @@
+        original_encode_queries = self._model.encode_queries
+        original_encode_corpus = self._model.encode_corpus
+
+        def wrapped_encode_queries(
```
I believe we should remove these

orionw replied:
Sure - but does this not cause issues with custom models?

KennethEnevoldsen replied:
I believe most custom models (if not all) have been refactored.

Samoed replied:
If you about models inside models dir, I changed them all

KennethEnevoldsen commented on mteb/models/cache_wrapper.py:
```diff
@@ -0,0 +1,330 @@
+
+        self._model.encode = wrapped_encode
+
+    def _cached_encode(
```
I would just call this `encode`

orionw replied:
I think that could get confusing with the models `encode` and this encode which is just a wrapper. Is this style or was there a particular other reason? If just style, I think I'd prefer to leave it.

KennethEnevoldsen commented on mteb/models/cache_wrapper.py:
```diff
@@ -0,0 +1,330 @@
+        logger.info(f"Closed TextVectorMap in directory: {self.directory}")
+
+
+class CachedEmbeddingWrapper:
```
```suggestion
class CachedEmbeddingWrapper(Encoder):
```

Samoed replied:
I've also created a dummy class for wrappers to indicate that the model is already wrapped. I think any new wrapper classes should inherit from this as well.

KennethEnevoldsen replied:
Will you link to it

Samoed replied:
https://github.com/embeddings-benchmark/mteb/blob/main/mteb/models/wrapper.py

orionw replied:
Gotcha, so inherit from Wrapper, Encoder then!

isaac-chung commented:
Sorry I was late to the party. I wish I was able to share some feedback sooner, esp. how this would interact with the user. I found @orionw 's original proposal to use this to be very convenient for the user. This could even be toggled by a flag with a default directory location, and the `path_to_cache_dir` param could be a secondary way to specific where if the user desires to.
```
evaluation.run(model, encode_kwargs={'cache_embeddings': 'path_to_cache_dir'})
```

The reason for me leaning towards the above option is that it does not require importing anything new from the existing imports. The only addition would be a flag (in CLI or the run method). 

That's said, I'm also just happy that we would now have this capability. If we go with this PR as is, the PR description can be updated to include the following:
```python
# define your task and model above as normal
...
# wrap the model with the cache wrapper
from mteb.models.cache_wrapper import CachedEmbeddingWrapper
model_with_cached_emb = CachedEmbeddingWrapper(model, cache_path='path_to_cache_dir')
# run as normal
evaluation.run(model, ...) 
```

orionw commented:
Thanks @isaac-chung! I am conflicted - I do think it's easier for the user (especially for CLI) as you mentioned and as I originally had, but I also agree with @KennethEnevoldsen that it's easier to maintain as a separate thing as we don't have to fit it into the MTEB class. 

I'm not sure how used this feature will be, but if it's a niche thing then it's fine to have these additional steps. Perhaps we leave it as is and can refactor into in the main code if we find it has a high usage/demand?

Thanks for the updated description for the PR, updated!

orionw commented on mteb/models/overview.py:
```diff
@@ -21,6 +21,8 @@
     nomic_models,
     openai_models,
+    promptriever_models,
+    repllama_models,
```
I was using these as a test and noticed they weren't here. Adding them but lmk if they were supposed to be elsewhere.

KennethEnevoldsen replied:
This is perfect

KennethEnevoldsen commented:
I believe this can be merged in once the conflict is resolved
